contracts/exchangeIssuance/ExchangeIssuance.sol

```diff
@@ -0,0 +1,768 @@
+            totalEth = totalEth.add(amountEth);
+        }
+        
+        if(address(_inputToken) == WETH) {
+            return totalEth;
+        }
+        
+        (uint256 tokenAmount, ) = _getMinTokenForExactToken(totalEth, address(_inputToken), address(WETH));
+        return tokenAmount;
+    }
+    
+    /**
+     * Returns an estimated amount of ETH or specified ERC20 received for a given SetToken and SetToken amount. 
+     *
+     * @param _setToken             SetToken redeemed
+     * @param _amountSetToRedeem    Amount of SetToken
+     * @param _outputToken          Address of output token. Ignored if _isOutputETH is true
+     * @return                      Estimated amount of ether/erc20 that will be received
+     */
+    function getEstimatedRedeemSetAmount(
+        ISetToken _setToken,
+        address _outputToken,
+        uint256 _amountSetToRedeem
+    ) 
+        isSetToken(_setToken)
+        external
+        view
+        returns (uint256)
+    {
+        require(_amountSetToRedeem > 0, "ExchangeIssuance: INVALID INPUTS");
+        
+        ISetToken.Position[] memory positions = _setToken.getPositions();
+        uint256 totalEth = 0;
+        for (uint256 i = 0; i < positions.length; i++) {
+            uint256 amount = uint256(positions[i].unit).mul(_amountSetToRedeem).div(1 ether);
+            
+            // get maximum amount of ETH received for a given amount of SetToken component
+            (uint256 amountEth, ) = _getMaxTokenForExactToken(amount, positions[i].component, WETH);
+            totalEth = totalEth.add(amountEth);
+        }
+        if(_outputToken == WETH) {
+            return totalEth;
+        }
+        
+        // get maximum amount of tokens for totalEth amount of ETH
+        (uint256 tokenAmount, ) = _getMaxTokenForExactToken(totalEth, WETH, _outputToken);
+        return tokenAmount;
+    }
+    
+    
+    /* ============ Internal Functions ============ */
+
+    /**
+     * Sets a max aproval limit for an ERC20 token, provided the current allowance is zero. 
+     * 
+     * @param _token    Token to approve
+     * @param _spender  Spender address to approve
+     */
+    function _safeApprove(IERC20 _token, address _spender) internal returns (uint256) {
+        if (_token.allowance(address(this), _spender) == 0) {
+            _token.safeIncreaseAllowance(_spender, PreciseUnitMath.maxUint256());
+        }
+    }
+    
+    /**
+     * Sells the total balance that the contract holds of each component of the set
+     * using the best quoted price from either Uniswap or Sushiswap
+     * 
+     * @param _setToken     The SetToken that is being liquidated
+     * @return              Amount of WETH received after liquidating all components of the SetToken
+     */
+    function _liquidateComponentsForWETH(ISetToken _setToken) internal returns (uint256) {
+        uint256 sumEth = 0;
+        ISetToken.Position[] memory positions = _setToken.getPositions();
```

Comment:
`IMPORTANT`: Use `.getComponents` vs. getPositions. If there are multiple positions with the same component, this would be buggy as the units of the components will be tried to be purchased multiple times

Please add a unit test where a SetToken has an external position (you can use something like the stake module)

Reply:
ExchangeIssuance should not handle token cases where there are external positions, so we should just grab the component and its units